eva-remapping-ingest/src/main/java/uk/ac/ebi/eva/remapping/ingest/batch/io/RemappedSubmittedVariantsWriter.java

```diff
@@ -17,58 +17,327 @@
+
+        List<SubmittedVariantEntity> svesToDiscard = Stream.concat(duplicateHashDiscards.getLeft().stream(),
+                                                                    duplicateAccessionDiscards.getLeft().stream())
+                                                            .collect(Collectors.toList());
+        List<SubmittedVariantOperationEntity> discardOperations = Stream.concat(duplicateHashDiscards.getRight().stream(),
+                                                                                duplicateAccessionDiscards.getRight().stream())
+                                                                        .collect(Collectors.toList());
+
+        BulkOperations bulkOperations = mongoTemplate.bulkOps(BulkOperations.BulkMode.UNORDERED,
+                                                              SubmittedVariantEntity.class,
+                                                              collection);
+        // Deal with discards before inserts, to avoid DuplicateKeyExceptions
+        if (svesToDiscard.size() > 0) {
+            bulkOperations.remove(query(where("_id").in(
+                    svesToDiscard.stream().map(SubmittedVariantEntity::getHashedMessage)
+                                 .collect(Collectors.toSet()))));
+            bulkOperations.execute();
+        }
+        if (discardOperations.size() > 0) {
+            try {
+                BulkOperations svoeBulkOperations = mongoTemplate.bulkOps(BulkOperations.BulkMode.UNORDERED,
+                                                                          SubmittedVariantOperationEntity.class,
+                                                                          operationCollection);
+                svoeBulkOperations.insert(discardOperations);
+                BulkWriteResult bulkWriteResult = svoeBulkOperations.execute();
+                remappingIngestCounts.addRemappedVariantsDiscarded(bulkWriteResult.getInsertedCount());
+
+            } catch (DuplicateKeyException exception) {
+                // As we check for hash collisions in submitted variants, we should only get duplicate keys from
+                // trying to insert an identical DISCARD operation, which should only happen when resuming or rerunning.
+                MongoBulkWriteException writeException = ((MongoBulkWriteException) exception.getCause());
+                BulkWriteResult bulkWriteResult = writeException.getWriteResult();
+
+                List<String> duplicatesSkipped = MongoBulkWriteExceptionUtils
+                        .extractUniqueHashesForDuplicateKeyError(writeException).collect(Collectors.toList());
+                logger.warn("Duplicate key exception when inserting DISCARD operations: " + duplicatesSkipped);
+
+                remappingIngestCounts.addRemappedVariantsDiscarded(bulkWriteResult.getInsertedCount());
+                remappingIngestCounts.addRemappedVariantsSkipped(duplicatesSkipped.size());
+            }
+        }
+
```

Comment:
I would suggest putting this first part of the in a function to be left with 
```java
svesToInsert = resolveDuplicate(svesToInsert);
if (svesToInsert.size() > 0) {
    bulkOperations.insert(svesToInsert);
...
```

Reply:
If I understand what you mean by this, then the `resolveDuplicates` function would have to take care of the discards and operations as well... for me this seems harder to follow than keeping all the database modifications in the write method, so I'm inclined to keep it as-is for now.